cql3/statements/delete_statement.cc

```diff
@@ -103,13 +162,12 @@ delete_statement::prepare_internal(database& db, schema_ptr schema, variable_spe
     }
     prepare_conditions(db, *schema, bound_names, *stmt);
     stmt->process_where_clause(db, _where_clause, bound_names);
-    if (!db.supports_infinite_bound_range_deletions()) {
-        if (!stmt->restrictions().get_clustering_columns_restrictions()->has_bound(bound::START)
-                || !stmt->restrictions().get_clustering_columns_restrictions()->has_bound(bound::END)) {
-            throw exceptions::invalid_request_exception("A range deletion operation needs to specify both bounds for clusters without sstable mc format support");
-        }
+    if (!db.supports_infinite_bound_range_deletions() &&
+        !bounded_ck(stmt->restrictions().get_clustering_columns_restrictions()->expression)) {
+        throw exceptions::invalid_request_exception(
+                "A range deletion operation needs to specify both bounds for clusters without sstable mc format support");
     }
-    if (stmt->restrictions().get_clustering_columns_restrictions()->is_slice()) {
+    if (has_slice(stmt->restrictions().get_clustering_columns_restrictions()->expression)) {
```

Comment:
Are has_slice and is_slice exactly equivalent? has_slice implies that `a>4 and a<9 and c LIKE '%664%'` will return true, while is_slice implies false.

Reply:
Current code saves itself by refusing filtering without ALLOW FILTERING, and not recognizing ALLOW FILTERING in delete statements. But we may have a bug here.

Reproducer (correctly rejected on master):

```
cassandra@cqlsh> create KEYSPACE ks1 WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1 };
cassandra@cqlsh> use ks1;
cassandra@cqlsh:ks1> create table tab (id int, ck int, r text, primary key(id, ck));
cassandra@cqlsh:ks1> delete from tab where id = 3 and ck > 5 and ck < 9 and r like '%a%';
InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot execute this query as it might involve data filtering and thus may have unpredictable performance. If you want to execute this query despite the performance unpredictability, use ALLOW FILTERING"
cassandra@cqlsh:ks1> delete from tab where id = 3 and ck > 5 and ck < 9 and r like '%a%' allow filtering;
SyntaxException: line 1:68  : syntax error...
```

Reply:
> Are has_slice and is_slice exactly equivalent? has_slice implies that `a>4 and a<9 and c LIKE '%664%'` will return true, while is_slice implies false.

Note that this is invoked on `.get_clustering_columns_restrictions()`, so your example is either [disallowed](https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/restrictions/single_column_primary_key_restrictions.hh?subtree=true#L142:140) (if both `a` and `c` are clustering columns) or `has_slice` is fully equivalent to `is_slice` (if they are not).

> But we may have a bug here.

I highly doubt it, based on what the `is_slice` definition was in the old code:
https://github.com/scylladb/scylla/blob/0e86f1bf6665fd9ce9c9b1f2577ed43b16f49f5f/cql3/restrictions/restriction.hh#L96-L98

Combined with this line from old code, it looks equivalent to me (apart from one special case I'll discuss below):
https://github.com/scylladb/scylla/blob/0e86f1bf6665fd9ce9c9b1f2577ed43b16f49f5f/cql3/restrictions/single_column_primary_key_restrictions.hh#L156

The special case is `ck1 < 2 AND ck1 = 0`.  This was previously rejected upstream because the old code couldn't represent it with its class hierarchy.  Now it is allowed, so it's subject to the check we're discussing here.  This check may reject it (for compact schemas or non-* column sets), even though it isn't _really_ a range deletion.  If you think that's a problem, we should discuss what should happen instead (and cover cases other than `= 0` for the second atom).